src/basicfuns.jl

```diff
@@ -246,6 +246,45 @@ function logsumexp(X::AbstractArray{T}; dims=:) where {T<:Real}
     end
 end
 
+"""
+    logsumexp_onepass(X)
+
+Compute `log(sum(exp, X))` in a numerically stable way that avoids intermediate under- and
+overflow.
+
+In contrast to [`logsumexp`](@ref) the result is computed using a single pass over the data.
+`X` should be an iterator of real numbers.
+
+# References
+
+[Sebastian Nowozin: Streaming Log-sum-exp Computation.](http://www.nowozin.net/sebastian/blog/streaming-log-sum-exp-computation.html)
+"""
+function logsumexp_onepass(X)
+    isempty(X) && return log(sum(X))
+
+    # initialize maximum value and accumulated sum for the first iterate
+    x, state = iterate(X)
+    xmax = x
+    r = exp(zero(x))
+    r_one = r
+
+    # for all other iterates
+    while (next = iterate(X, state)) !== nothing
```

Comment:
FWIW, if you're looking for the last bit of performance, it could be faster to have another method specific to `AbstractArray` using `@inbounds @simd for x in X`. Often it makes a difference.